src/diagram/join.rs

```diff
@@ -0,0 +1,322 @@
+use std::any::TypeId;
+
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use smallvec::SmallVec;
+use tracing::debug;
+
+use crate::{Builder, IterBufferable, Output};
+
+use super::{
+    DiagramError, DynOutput, NextOperation, NodeRegistry, SerializeMessage, SourceOperation,
+};
+
+#[derive(Debug, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct JoinOp {
+    pub(super) next: NextOperation,
+
+    /// Controls the order of the resulting join. Each item must be an operation id of one of the
+    /// incoming outputs.
+    pub(super) inputs: Vec<SourceOperation>,
+
+    /// Do not serialize before performing the join. If true, joins can only be done
+    /// on outputs of the same type.
+    pub(super) no_serialize: Option<bool>,
+}
+
+pub(super) fn register_join_impl<T, Serializer>(registry: &mut NodeRegistry)
+where
+    T: Send + Sync + 'static,
+    Serializer: SerializeMessage<Vec<T>>,
+{
+    if registry.join_impls.contains_key(&TypeId::of::<T>()) {
+        return;
+    }
+
+    registry
+        .join_impls
+        .insert(TypeId::of::<T>(), Box::new(join_impl::<T>));
+}
+
+/// Serialize the outputs before joining them, and convert the resulting joined output into a
+/// [`serde_json::Value`].
+pub(super) fn serialize_and_join(
+    builder: &mut Builder,
+    registry: &NodeRegistry,
+    outputs: Vec<DynOutput>,
+) -> Result<Output<serde_json::Value>, DiagramError> {
+    debug!("serialize and join outputs {:?}", outputs);
+
+    if outputs.is_empty() {
+        // do not allow empty joins
+        return Err(DiagramError::EmptyJoin);
+    }
+
+    let outputs = outputs
+        .into_iter()
+        .map(|o| {
+            let serialize_impl = registry
+                .serialize_impls
+                .get(&o.type_id)
+                .ok_or(DiagramError::NotSerializable)?;
+            let serialized_output = serialize_impl(builder, o)?;
+            Ok(serialized_output)
+        })
+        .collect::<Result<Vec<_>, DiagramError>>()?;
+
+    // we need to convert the joined output to [`serde_json::Value`] in order for it to be
+    // serializable.
+    let joined_output = outputs.join_vec::<4>(builder).output();
+    let json_output = joined_output
+        .chain(builder)
+        .map_block(|o| serde_json::to_value(o).unwrap())
```

Comment:
What are the conditions in which this `.unwrap()` might panic? We should either document a rationale for why a panic is impossible, or else we should use `.cancel_on_err()` here instead of `.unwrap()`.

Reply:
Technically this will never fail because we are converting a (Small)Vec of Value to a Value. But I made it `cancel_on_err` just in case.